### DISCUSSION TITLE
A Custom Implementation of the ByteStream Driver Model to support a Network of Deployments

### DISCUSSION THREAD

**Discussion by Emil808** [nasa/fprime#1469] [Show and tell]
I have been working on an implementation for the ByteStream Driver Model to support a Network of Deployments. I built some classes inheriting the TCP Server and TCP Client classes that were present, and based heavily on the operations of the Drv/Ip files. 

https://github.com/Emil808/fprime/blob/multi_tcp/DepA/MultiNode.md

The goal was to allow communication between identical Deployments/Space Vehicles, working together as a swarm. For the example I present here, each Deployment has their own GPS position (in ECEF), and is sending it out to the other nodes. When receiving another Node's GPS position, the receiving Node will calculate the distance between itself and the other Node. 

Suggestions, and criticism would be appreciated to help me get better at this. Thank you!

**Comment by ThibFrgsGmz** [comment]
I haven't run it yet but it looks amazing!

As far as managing the configuration of your project, you could use fprime as a sub-module as shown in [separated-project-example](https://github.com/fprime-community/separated-project-example).

This would make it easier to see which components you had to create specifically for your mission. And you could more easily manage your dependency on fprime by simply updating the hash of the submodule.

**Reply by Emil808** [reply]
I made an additional component within the Drv/ directory, so how would I integrate that if the rest of the Fprime code is a submodule?

**Reply by ThibFrgsGmz** [reply]
You can take a look on how it is made in https://github.com/fprime-community/fprime-arduino
You play with the cmake to include your additional components.

**Reply by capsulecorplab** [reply]
@ThibFrgsGmz do you have a working fork of `fprime-arduino`, out of curiosity? I'm constantly running into CMake errors while trying to generate the build cache

**Reply by ThibFrgsGmz** [reply]
I don't use this repo directly for my project management. For this way of managing FPrime I am mostly inspired by [fprime-sphinx](https://github.com/fprime-community/fprime-sphinx) or [fprime-azure-threadx](https://github.com/fprime-community/fprime-azure-threadx/tree/fprime-azure-threadx-cmjl-14-01-2022-v00).

What problems do you encounter on CMake ?

**Reply by capsulecorplab** [reply]
https://github.com/fprime-community/fprime-arduino/issues/7 for the fprime v1.5.0 submodule. Otherwise, I'm encountering issues with prescan on v3.0.0 or `devel` branch

**Reply by SterlingPeet** [reply]
I have tracked down the bug in fprime-arduino that is preventing it from working with more modern F Prime, but I haven't had time to package the issue and and solution for presentation to the F Prime developers.  It is an issue with the type-handling system, so Its not a super quick patch (at least, not an obvious one at my first glance).

**Comment by Emil808** [comment]
Update! 

New Repository: https://github.com/Emil808/fprime-Mesh_Network

I separated the deployment and new driver files, with fprime as a sub-module. Should work with ver. 3.0.0

**Reply by ThibFrgsGmz** [reply]
It's better this way!

Concerning the [telemetry data](https://github.com/Emil808/fprime-Mesh_Network/blob/2909f5b2f1ae7a206ae782786f50631825d0fc6e/DepA/sGPS/sGPS.fpp#L67:74) of your GPS component, you didn't group everything in one data structure because the GDS doesn't manage the display of complex data structures?

**Reply by Joshua-Anderson** [reply]
Super impressive! Very cool demonstration of mesh networking.

**Reply by Emil808** [reply]
@ThibFrgsGmz 
Regarding the telemetry, I went through [the fpp guide](https://fprime-community.github.io/fpp/fpp-users-guide.html#Defining-Components_Telemetry_Identifiers) and found that telemetry does support arrays.
